File: 48-rotate-image/48-rotate-image.go
func rotate(matrix [][]int)  {
    n := len(matrix) - 1
    
    swap := func (i, j int) {
        matrix[i][j], matrix[j][n - i], matrix[n - i][n - j], matrix[n - j][i] = matrix[n - j][i], matrix[i][j], matrix[j][n - i], matrix[n - i][n - j]
    }
    
    for i := 0; i < n; i++ {
        for j := i; j < n - i; j++ {
            swap(i, j)
        }
    }
}

File: 48-rotate-image/48-rotate-image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
  const n = matrix.length - 1;
  
  const swap = function (i, j) {
    [
      matrix[i][j],
      matrix[j][n - i],
      matrix[n - i][n - j],
      matrix[n - j][i],
    ] = [
      matrix[n - j][i],
      matrix[i][j],
      matrix[j][n - i],
      matrix[n - i][n - j],
    ]
  };
  
  for (let i = 0; i < n; i++) {
    for (let j = i; j < n - i; j++) {
      swap(i, j);
    }
  }
};
